### DISCUSSION TITLE
esp32 esp-idf v4.4.4 mongoose.ws 7.10 crashes during websocket update

### DISCUSSION THREAD

**Discussion by steff2000** [cesanta/mongoose#2232] [Ask questions, get answers] [answered]
Hi folks,
my mongoose webserver appllication updates clients via websocket, so far so good.
Sometimes the application crashes with a Guru Meditation Error. 

As I see it, the error only happens when another websocket client connects.
I think I have seen the error also when a disconnect (web browser window closed) occurs.

Guru Meditation Error: Core  1 panic'ed (LoadStoreError). Exception was unhandled.

Core  1 register dump:
PC      : 0x4000c3f5  PS      : 0x00060f30  A0      : 0x800e5efc  A1      : 0x3ffd7fd0  
A2      : 0x3ffe8c71  A3      : 0x3ffe8c67  A4      : 0xfffffded  A5      : 0x0000000a  
A6      : 0x3ffce604  A7      : 0x00000800  A8      : 0x0001738f  A9      : 0x40000000  
A10     : 0x00000000  A11     : 0x295a3860  A12     : 0x00000000  A13     : 0x3ffd67d0  
A14     : 0x00000003  A15     : 0x00060023  SAR     : 0x00000020  EXCCAUSE: 0x00000003  
EXCVADDR: 0x40000000  LBEG    : 0x400014fd  LEND    : 0x4000150d  LCOUNT  : 0xffffffdf  

Backtrace: 0x4000c3f2:0x3ffd7fd0 0x400e5ef9:0x3ffd7ff0 0x400e5f52:0x3ffd8030 0x400e5f9a:0x3ffd8060 0x400db55b:0x3ffd80b0 0x40090f6a:0x3ffd8130
0x400e5ef9: mg_ws_wrap at /home/user/Dokumente/esp32_projekte/websocket_mqtt/main/mongoose.c:6325

0x400e5f52: mg_ws_vprintf at /home/user/Dokumente/esp32_projekte/websocket_mqtt/main/mongoose.c:6056

0x400e5f9a: mg_ws_printf at /home/user/Dokumente/esp32_projekte/websocket_mqtt/main/mongoose.c:6064

0x400db55b: WebSocketTask at /home/user/Dokumente/esp32_projekte/websocket_mqtt/main/WebSocketTask.c:98

0x40090f6a: vPortTaskWrapper at /home/user/Dokumente/esp-idf/components/freertos/port/xtensa/port.c:142

WebSocketTask,c:

                        [snip]
		        char *out = cJSON_Print(root);
			cJSON_Delete(root);
			for (c = mgr_p->conns; c != NULL; c = c->next) {
				if (c->data[0] != 'W') {
					continue;
				}
L98				size_t send = mg_ws_printf(c, WEBSOCKET_OP_TEXT, "%s", out);
			} // for (c = mgr->conns; c != NULL; c = c->next)
			free(out);
                        [snip]

**Comment by scaprile** [comment, marked as answer]
I can't guess what you are doing, nevertheless the word 'Task' above sends shivers down my spine.

From [the docs](https://mongoose.ws/documentation/#2-minute-integration-guide)
> NOTE: Since Mongoose's core is not protected against concurrent accesses, make sure that all mg_* API functions are called from the same thread or RTOS task.

If you absolutely need multi-threading, take the time to go through the user guide (linked above) and [tutorials](https://mongoose.ws/tutorials/), there is one that shows you how to do [multi-threading](https://mongoose.ws/tutorials/multi-threaded/).

If my guessing was wrong, please open an issue, fill the issue template, and provide the minimum piece of code to reproduce your issue.

**Comment by cpq** [comment]
@steff2000 Note: mongoose's *printf functions can print JSON. You don't need cJSON or any other JSON library.
